### PR TITLE
Plumb PouchDB persistence into pipes-shell

### PR DESCRIPTION
**Pull request by sjmiles**
- create infrastructure to support both node and web versions of pipe-shell
- simplify deployment
  - reduced to two js files (minimal, one for main thread one for worker)
  - eliminated some unneeded particles/* files
- support persistent context storage for pipe-entities
  - uses PouchDB on WebView
  - uses Volatile storage on Nodejs

### CODE REVIEW

File: particles/PipeApps/source/RecentAddresses.js [hunk 1]
```diff
@@ -15,25 +15,17 @@ defineParticle(({DomParticle, html, log}) => {
...
-          name: address.rawData.address,
-          source: 'com.weaseldev.fortunecookies'
-        };
-        return JSON.stringify(result);
-      }
+    query(entities) {
+      const addresses = entities.filter(entity => entity.rawData.type === 'address');
+      const sorted = addresses.sort((a, b) => (b.rawData.timestamp || 0) - (a.rawData.timestamp || 0));
+      const result = sorted[0] || Object;
+      return JSON.stringify(result.rawData);
```
Comment (shaper) [inline particles/PipeApps/source/RecentAddresses.js]: If we end up with just `Object` won't this line fail dereferencing `rawData`.

File: particles/PipeApps/MapsApp.recipes [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
 schema Address
   Text address
 
 particle RecentAddresses in './source/RecentAddresses.js'
-  in [Address] recentAddresses
+  in [PipeEntity] recentEntities
   out Address address
 
 recipe RecentAddresses
+  map 'pipe-entities' as recentEntities
```
Comment (shaper) [inline particles/PipeApps/MapsApp.recipes]: Does the 'pipe-entities' thing work here? Was it not working in a `create` the other day when we were looking at it together?

Ah, I see you have a comment later that maybe is about this. I couldn't find an existing issue for this syntax, can you file one with the example?

File: shells/pipes-shell/context.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    dumpStores([this.entityStore]);
+  }
+  async getInitialChange(store) {
+    const change = {add: []};
+    const values = await store.toList();
+    values.forEach(value => change.add.push({value}));
+    return change;
+  }
+  async entityStoreChange(change) {
+    console.log(change);
```
Comment (shaper) [inline shells/pipes-shell/context.js]: Could be overly verbose logging?

File: shells/pipes-shell/pipe.js [hunk 4]
```diff
@@ -106,10 +133,21 @@ const instantiateRecipe = async (arc, manifest, name) => {
   }
 };
 
+const oneChange = (store, callback, arc) => {
```
Comment (shaper) [inline shells/pipes-shell/pipe.js]: It's easy to mis-read `oneChange` and `onChange`, could be worth a more differentiated name for one or the other.

File: shells/web-shell/elements/pipes/device-client-pipe.js [hunk 5]
```diff
@@ -112,10 +112,12 @@ class DeviceClientPipe extends Xen.Debug(Xen.Async, log) {
   update({userid, context, suggestions, arc, pipearc}, state) {
     if (pipearc && !state.pipeStore) {
       state.pipeStore = pipearc._stores[0];
-      log('got pipeStore', state.pipeStore);
-      // retry
       if (!state.pipeStore) {
+        // retry
+        log('looking for pipestore...');
```
Comment (shaper) [inline shells/web-shell/elements/pipes/device-client-pipe.js]: If there isn't already one, can you file an issue for Runtime team to request a hook we can watch for once stores are loaded?

File: shells/pipes-shell/node/node.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+global.ShellApi = DeviceApiFactory(`volatile://`, global.DeviceClient);
+
+console.log(`version: feb-27.0`);
+
+// configure arcs environment
+Utils.init(global.envPaths.root, global.envPaths.map);
+
+// test it
+
+setTimeout(() => {
```
Comment (shaper) [inline shells/pipes-shell/node/node.js]: Maybe comment this out for checkin?

File: particles/PipeApps/source/RecentAddresses.js (hunk 1 above)
Comment (sjmiles) [reply particles/PipeApps/source/RecentAddresses.js]: No, trying to deref a property off of a non-object (like null) throws an error, but dereferencing an undefined property on any object just returns undefined. `Object` makes a convenient null-object in this scenario (and it's slightly faster than `{}`);

File: particles/PipeApps/MapsApp.recipes (hunk 2 above)
Comment (sjmiles) [reply particles/PipeApps/MapsApp.recipes]: This version of the store is created in code, where I can easily set the `id`. It's the manifest syntax that doesn't seem to work.

The example was my broken project which I worked-around, so I don't have a convenient repro. This happens a lot (aka trying to capture a repro can be time-consuming and feels off-task). But I will file a ticket with the info.

File: shells/pipes-shell/node/node.js (hunk 6 above)
Comment (sjmiles) [reply shells/pipes-shell/node/node.js]: Ok. 

Fwiw, I don't expect this version (NodeJs target) of the project works in practice yet. We have yet to run our first Nodejs-on-Android test. I built this mostly so I could compare the dependency trees of Arcs on web vs nodejs, and also just to keep myself honest about making dual-targets work.

File: shells/web-shell/elements/pipes/device-client-pipe.js (hunk 5 above)
Comment (sjmiles) [reply shells/web-shell/elements/pipes/device-client-pipe.js]: I'll look into it. 

Fwiw, the `web-shell` version of `device-client-pipe` is now deprecated for `pipe-shell`, so this entire module is about to go away.

File: shells/pipes-shell/pipe.js (hunk 4 above)
Comment (sjmiles) [reply shells/pipes-shell/pipe.js]: Yes, I saw that and I should have fixed it. 

Fwiw, this code is going to undergo a fair amount of refactoring in the near future and I instinctively don't spend time on things I know will be better abstracted soon.

File: shells/pipes-shell/context.js (hunk 3 above)
Comment (sjmiles) [reply shells/pipes-shell/context.js]: Hard to say, logging becomes helpful when problems occur. Until we have some successful roundtrips on actual device, I don't know what kind of information we will need. I tried to limit myself to critical junctures; this spot relies on a runtime callback to propagate observedEntities into the context.

File: particles/PipeApps/source/RecentAddresses.js (hunk 1 above)
Comment (sjmiles) [reply particles/PipeApps/source/RecentAddresses.js]: Just to verify
image

**Pull request merged**